ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java

```diff
@@ -30,22 +34,59 @@
 
 public class ForkIdManager {
 
-  private final Blockchain blockchain;
   private final Hash genesisHash;
-  private final List<Long> forks;
-  private long forkNext;
-  private final long highestKnownFork;
-  private List<ForkId> forkAndHashList;
+  private final List<ForkId> forkAndHashList;
+
+  private final List<ForkIDChecker> forkIDCheckers;
 
   public ForkIdManager(final Blockchain blockchain, final List<Long> forks) {
-    this.blockchain = blockchain;
+    assert blockchain != null && forks != null;
     this.genesisHash = blockchain.getGenesisBlock().getHash();
-    // de-dupe and sanitize forks
-    this.forks =
-        forks.stream().filter(fork -> fork > 0).distinct().collect(Collectors.toUnmodifiableList());
-    highestKnownFork = forks.size() > 0 ? forks.get(forks.size() - 1) : 0L;
-    createForkIds();
-  };
+    this.forkAndHashList = new ArrayList<>();
+    final ForkIDChecker legacyForkIdChecker =
+        createForkIDChecker(
+            blockchain,
+            genesisHash,
+            forks,
+            fs ->
+                fs.stream()
+                    .filter(fork -> fork > 0)
+                    .distinct()
+                    .collect(Collectors.toUnmodifiableList()),
+            forkAndHashList);
+    // if the fork list contains only zeros then we may be in a consortium/dev network
+    if (onlyZerosForkBlocks(forks)) {
```

Comment:
What happens if we have a few 0 forks but others greater than 0? Would we do the right thing?

Reply:
Yes we would to the right thing. This is what happens on Goerli:
```json
{
  "config":{
    "chainId":5,
    "homesteadBlock":0,
    "eip150Block":0,
    "eip150Hash": "0xbf7e331f7f7c1dd2e05159666b3bf8bc7a8a3a9eb1d518969eab529dd9b88c1a",
    "eip155Block":0,
    "eip158Block":0,
    "eip160Block":0,
    "byzantiumBlock":0,
    "constantinopleBlock":0,
    "petersburgBlock":0,
    "istanbulBlock":1561651,
    "clique":{
      "blockperiodseconds":15,
      "epochlength":30000
    }
  }
```

Reply:
Unit tests and the EIP2124 spec also validate that the correct hashes are arrived at.